nonmember_subscript_operator/main.tex

```diff
@@ -125,6 +126,68 @@ \subsection{Compatiblity layer}
+This issue could be worked around using a custom type which cannot truncate into an integral, allowing to e.g. also specify the used interpolation:
+\smallskip\begin{lstlisting}
+std::vector v{0.0, 1.0, 4.5, 7.8};
+double value = v[1.6_linear]; // value == 3.1
+\end{lstlisting}
+
+With the adoption of P2128, we could even allow interpolation on higher dimensional objects, like the proposed `std::mdspan` or `std::mdarray`.
+\smallskip\begin{lstlisting}
+std::mdarray<double, N, M, K> a = ...;
+double value = a[7.5, 8.9, 76.4]; // trilinear interpolation
+\end{lstlisting}
+
+While we are aware that such a usage might be niche or a separate function for interpolated access could be arguably better, we would like to give library writes the necessary machinery to implement such a feature within their domains if they desire so.
+
+\subsection{Generic subsetting}
+The idea from section \ref{sec:simd_load} to use SIMD vectors as subscripts can be further generalized to allow any kind entity representing multiple indices as subscript.
+\smallskip\begin{lstlisting}
+auto operator[](std::ranges::random_access_range r, std::ranges::input_range indices);
+\end{listing}
+The concrete implementation and return type of such an operation is debatable and thus intentionally omitted.
+Still, it could allow for better formulation of some algorithms or more efficient access for certain ranges, if the index set used for subscription is known in advance.
+
+\subsection{Predicated subsetting}
+Subscripting with a unary predicate would allow us to select a subset of an existing container.
+Combined with an easy way to specify these predicates, e.g. Boost.Lambda2, this allows powerful and concise expressions:
+\smallskip\begin{lstlisting}
+template <std::ranges::random_access_range R, typename UnaryPredicate>
+auto operator[](R&& r, UnaryPredicate&& p) {
+	return std::forward<R>(r) | std::views::filter(std::forward<UnaryPredicate>(p));
+}
+std::vector v{0.0, -1.0, 4.5, -7.8};
+for (auto e : v[_1 > 0])
+	...;
+\end{listing}
+
+\subsection{Slicing}
+With the adoption of P2128, a generic slicing facility could be defined:
+\smallskip\begin{lstlisting}
+template <std::ranges::random_access_range R>
+auto operator[](R r, std::size_t from, std::size_t to, std::size_t step = 0) { ... }
+
+std::vector v{...};
+auto slice1 = v[10, 20];    // elements from the 10th to the 20th (exclusive)
+auto slice2 = v[10, 20, 2]; // every second element from the 10th to the 20th (exclusive)
+\end{listing}
+Such slicing operators are supported in other languages, e.g. Python.
+There have also been C++ language extensions to add such slicing functionality, e.g. Intel's Cilk Plus Array Notations.
+The adoption of P2128 and this proposal would allow for corresponding library solutions to emerge.
```

Comment:
Slicing ala Cilk+ would be quite nice for certain (numeric) domains. Generalizing the subscript operator enables emulating this without dedicated language support, but I'm not sure if that syntax would be desirable for such a feature...

Reply:
`v[10, 20, 2]` is indeed a bit ambiguous. It could also mean take the 10th, 20th and 2nd element. But then, should we disallow users to write that if they want? In Python, I can write `v[10:20:2]` and people usually understand what that means. Good conventions can establish in the communities.

Reply:
> v[10, 20, 2] is indeed a bit ambiguous. It could also mean take the 10th, 20th and 2nd element.
Without knowing the type of `v` it could be literally anything - e.g. for a matrix it could simply mean: v[10, 20..2].

> In Python, I can write v[10:20:2] and people usually understand what that means. 
Yes, but:
1. Does Python actually have multi-dim subscripting?
2. If yes: Does it use `:` for that purpose or is that a different syntax? (e.g. `,`)

It's beyond the scope of the paper but IMHO a multi-dim subscript operator is ill-suited for a slicing mechanic as it conflates indices with slices.

Reply:
From what I have seen so far, I am by far not fluent in Python and the popular Numpy, there is a distinction:
* Multidim array access is done with repeated subscripts: `m[4][6]`
* Slicing is done using a different notation: `m[4:6]`, meaning the two elements at index 4 and 5.
* The notation `m[4, 6]` apparently creates a tuple and passes that to the subscript operator, resulting in an error.

Reply:
> a multi-dim subscript operator is ill-suited for a slicing mechanic as it conflates indices with slices.

I guess you are right here as well. Slicing using `operator[]` is probably confusing.